app/controllers/routing/index/RemoveCountryOfRoutingYesNoController.scala

```diff
@@ -74,14 +80,48 @@ class RemoveCountryOfRoutingYesNoController @Inject() (
             formWithErrors => Future.successful(BadRequest(view(formWithErrors, lrn, mode, index, request.arg))),
             {
               case true =>
-                CountryOfRoutingSection(index)
-                  .removeFromUserAnswers()
-                  .updateTask()
-                  .writeToSession()
-                  .navigateTo(addAnother(lrn, mode))
+                for {
+                  updatedAnswers <- Future.fromTry(
+                    request.userAnswers
+                      .remove(CountryOfRoutingSection(index))
+                      .flatMap(findAndRemoveOffices(_, OfficesOfTransitSection, OfficeOfTransitPage, OfficeOfTransitSection))
+                      .flatMap(findAndRemoveOffices(_, OfficesOfExitSection, OfficeOfExitPage, OfficeOfExitSection))
+                  )
+                  _ <- sessionRepository.set(updatedAnswers)
+                } yield Redirect(addAnother(lrn, mode))
+
               case false =>
                 Future.successful(Redirect(addAnother(lrn, mode)))
             }
           )
     }
+
+  private def findAndRemoveOffices(
+    userAnswers: UserAnswers,
+    sections: Section[JsArray],
+    page: Index => QuestionPage[CustomsOffice],
+    section: Index => Section[JsObject]
+  )(implicit request: SpecificDataRequestProvider1[Country]#SpecificDataRequest[AnyContent]): Try[UserAnswers] = {
+
+    case class OfficeWithIndex(office: CustomsOffice, index: Index)
+
+    val officesWithIndex: Seq[OfficeWithIndex] = request.userAnswers
+      .get(sections)
+      .map(_.value.zipWithIndex.flatMap {
+        case (_, index) => request.userAnswers.get(page(Index(index))).map(OfficeWithIndex(_, Index(index)))
+      })
+      .getOrElse(Seq.empty[OfficeWithIndex])
+      .toSeq
+
+    val officesToDelete: Seq[Index] = officesWithIndex.collect {
+      case officeWithIndex if officeWithIndex.office.countryId == request.arg.code.code => officeWithIndex.index
+    }
+
+    officesToDelete.foldRight(Try(userAnswers)) {
+      case (index, acc) =>
+        acc.flatMap {
+          _.remove(section(index))
+        }
+    }
+  }
```

Comment:
Rather than iterating over the list to establish the indices, then iterating again to establish which ones to delete, then iterating again to do the deletions, why don't we just to do it all in one? Something like

```
  private def findAndRemoveOffices(
    userAnswers: UserAnswers,
    array: Section[JsArray],
    obj: Index => Section[JsObject],
    page: Index => QuestionPage[CustomsOffice]
  )(implicit request: Request): Try[UserAnswers] = {

    val numberOfOffices = userAnswers.get(array).length
    (0 until numberOfOffices).map(Index(_)).foldRight(Try(userAnswers)) {
      case (index, acc) =>
        userAnswers.get(page(index)) match {
          case Some(value) if value.countryId == request.arg.code.code => acc.flatMap(_.remove(obj(index)))
          case _                                                       => acc
        }
    }
  }
```

Reply:
Yep, that makes more sense. Implemented